### ISSUE TITLE
Turning URLs into links

### ISSUE THREAD

**Issue by mfenner** [inukshuk/jekyll-scholar#30]
I want URLs in citations to be links, and I have monkey patched the `reference_tag` method to allow this:

``` ruby
module Jekyll
  class Scholar
    module Utilities

      alias_method :original_reference_tag, :reference_tag

      def reference_tag(entry)
        return missing_reference unless entry

        # Change URLs into markdown links
        entry["url"] = "[#{entry["url"]}](#{entry["url"]})" if entry["url"]

        entry = entry.convert(*bibtex_filters) unless bibtex_filters.empty?
        reference = CiteProc.process entry.to_citeproc,
          :style => style, :locale => config['locale'], :format => 'html'

        content_tag reference_tagname, reference,
          :id => [prefix, entry.key].compact.join('-')
      end
    end
  end
end
```

You can see that I change `entry["url"]` into a markdown link, which is then picked up my the markdown processor. You can see this at work at http://blog.martinfenner.org/about.html. Is this the correct way of doing this? I'm using Pandoc as markdown processor and in the default configuration it doesn't recognize blank URLs.

We could do something similar for `entry["doi"]`, as some styles display the DOI.

**Comment by ralphvanetten** [comment]
I tried this and it just shows the markup in the bibliography list.

I have:
```
@article{test,
      author = {test},
      title = {test},
      URL = {http://google.com}
}
```

which shows as `[1]test, test, (n.d.). [http://google.com](http://google.com).` without converting the url into a clickable link.

Do I need to configure something else?

**Comment by inukshuk** [comment]
I think we never actually added the filter, so what you need to do is add it to your `_plugins/ext.rb` file:

    require 'jekyll/scholar'
    require 'uri'

    module MarkdownFilter
      class Markdown < BibTeX::Filter
        def apply(value)
          value.to_s.gsub(URI.regexp(['http','https','ftp'])) { |c| "[#{$&}](#{$&})" }
        end
      end
    end

And then enable it in `_config.yml`:

    scholar:
      bibtex_filters:
        - latex
        - markdown

**Comment by ralphvanetten** [comment]
yes, that is what I did

**Comment by inukshuk** [comment]
@mfenner is it OK if I add the plugin? Looks like I should write a test for it anyway : )

**Comment by inukshuk** [comment]
I think for this to work the way @mfenner intended this needs to happen inside a markdown file, because you need the markdown processor to then convert the link.

Alternatively, you could adjust the filter above to generate an HTML link tag instead. Something like:

    value.to_s.gsub(URI.regexp(['http','https','ftp'])) { |c| "<a href=\"#{$&}\">#{$&}</a>" }

Better yet, you could adjust your [bibliography template](https://github.com/inukshuk/jekyll-scholar#bibliography-template)

**Comment by mfenner** [comment]
@inukshuk fee free to add the plugin.

**Comment by ralphvanetten** [comment]
Changing the filter to generate a HTML link tag works.

Thanks.

**Comment by rriemann** [comment]
I tried to add the markdown filter to make it work in this commit of my blog: https://github.com/rriemann/blog.riemann.cc/commit/e6d937db5bf321db1c40901e4226fa47f9b889c7

My bib file has doi fields e.g. `doi = {10.1007/978-3-319-58469-0_12},` that get rendered as `https://doi.org/10.1007/978-3-319-58469-0_12` but that are still not clickable.

The page is online here: https://blog.riemann.cc/output/

What is missing? Should I use in the Gemfile the jekyll_plugins group?

**Comment by inukshuk** [comment]
The markdown filter is already part of Jekyll-Scholar so you don't need to add it yourself; you only need to enable it in your configuration.

What the filter does is convert URLs to markdown link syntax (these are later turned into HTML links when the markdown is being processed). This filter does not turn DOIs into URLs though. If that's happening on your page, it's probably another filter at work (which likely runs after the markdown filter, so those DOI URLs won't be detected). You could make sure your DOIs are converted before the markdown filter processes the values or have the DOI filter create markdown links in the first place.

**Comment by rriemann** [comment]
The CSL default converts DOIs to links `https://doi.org/10.1007/978-3-319-59665-5_3`. So I think links should just work if the markdown filter is activated.

**Comment by ashinkarov** [comment]
You can modify the csl file and add a piece of code that outputs dois in the preferred format.
See how I handle urls:
https://github.com/ashinkarov/ashinkarov.github.com/blob/source/plugins/ieee.csl#L180
You can do similar stuff for dois.  Then you just point bibproc to use this file.

**Comment by inukshuk** [comment]
When jekyll-scholar renders your reference, this is what happens:
1. The item is read from your BibTeX file
2. BibTeX filters are applied (e.g. URLs converted to Markdown syntax)
3. BibTeX item is converted to CSL JSON
4. Item is rendered using CSL Style
5. The reference and the BibTeX item are passed to the template
6. The template is further processed by Jekyll (e.g., Markdown turned into HTML)

In other words, if your DOI field is rendered as a URL by the CSL style this is something that the Markdown BibTeX filter would not touch, because it is applied to the values as they come out of your BibTeX file, not the rendered reference.

As @ashinkarov points out, one option is to adjust your CSL style to output the DOI as Markdown style link.

**Comment by inukshuk** [comment]
Another option would be to pass your rendered reference string through the markdown filter from within your template.

**Comment by nrenga** [comment]
How to exactly do this in the template?

**Comment by inukshuk** [comment]
The angle brackets are inserted into your markdown documents as HTML entities `&lt;` and `&gt;`. It would be easier to use `[]` instead.

**Comment by benswift** [comment]
Hey, thanks for taking the time to help out.

I tried this in the `apa.csl` file:

```
<text variable="DOI" prefix="[link](https://doi.org/" suffix=")"/>
```
and it renders out like (e.g.) this in the final `_site` folder:
```
<li><span id="swiftVisualCodeAnnotations2013">Swift, B., Sorensen, A., Gardner, H. J., &amp; Hosking, J. (2013). Visual Code Annotations for Cyberphysical Programming. In <i>2013 1st International Workshop on Live Programming (LIVE)</i> (pp. 27–30). [link](https://doi.org/10.1109/LIVE.2013.6617345)</span></li>
```
so it doesn't seem to be getting markdownified.

The "using `&lt;` and `&gt;`" thing was based on [this example](https://github.com/ashinkarov/ashinkarov.github.com/blob/source/plugins/ieee.csl#L180) linked in an earlier comment, but as I already mentioned that doesn't work either.

Or maybe I'm just doing it wrong.

**Comment by inukshuk** [comment]
OK I just tried this using the markdown filter above and I noticed two things: first, if you're using the `kramdown` Markdown engine, you need to enable parsing inside of HTML blocks (that's disabled by default); and secondly... the output of from the liquid tag does not seem to be run through markdown anymore, like you say. This is probably due to a change in Jekyll?

**Comment by benswift** [comment]
Yep, I tried `parse_block_html: true` and tried modified the `.csl` file to output both markdown and html links, no success either way.

It's not a dealbreaker---people can copy-paste it into their url bar 😉 So I think I'll park it for now. Thanks for the work you did to investigate.

**Comment by benswift** [comment]
If it's helpful for others who come across this, my blog source is up at <https://github.com/benswift/benswift.github.io>.

And here's the gory details from `bundle show`
```
  * activesupport (5.2.2)
  * addressable (2.5.2)
  * autoprefixer-rails (9.4.5)
  * bibtex-ruby (4.4.7)
  * bundler (1.17.2)
  * citeproc (1.0.9)
  * citeproc-ruby (1.1.10)
  * colorator (1.1.0)
  * colorize (0.8.1)
  * concurrent-ruby (1.1.4)
  * csl (1.5.0)
  * csl-styles (1.0.1.9)
  * em-websocket (0.5.1)
  * ethon (0.12.0)
  * eventmachine (1.2.7)
  * execjs (2.7.0)
  * extras (0.3.0)
  * fastimage (2.1.5)
  * ffi (1.10.0)
  * forwardable-extended (2.6.0)
  * html-proofer (3.10.0)
  * http_parser.rb (0.6.0)
  * i18n (0.9.5)
  * jekyll (3.8.5)
  * jekyll-assets (3.0.12)
  * jekyll-compose (0.9.0)
  * jekyll-feed (0.11.0)
  * jekyll-paginate-v2 (2.1.0)
  * jekyll-sanity (1.2.0)
  * jekyll-sass-converter (1.5.2)
  * jekyll-scholar (5.14.0)
  * jekyll-seo-tag (2.5.0)
  * jekyll-watch (2.1.2)
  * kramdown (1.17.0)
  * latex-decode (0.3.1)
  * liquid (4.0.1)
  * liquid-tag-parser (1.9.0)
  * listen (3.1.5)
  * mercenary (0.3.6)
  * mini_portile2 (2.4.0)
  * minitest (5.11.3)
  * namae (1.0.1)
  * nokogiri (1.10.1)
  * parallel (1.12.1)
  * pathutil (0.16.2)
  * public_suffix (3.0.3)
  * rack (2.0.6)
  * rb-fsevent (0.10.3)
  * rb-inotify (0.10.0)
  * rouge (3.3.0)
  * ruby_dep (1.5.0)
  * safe_yaml (1.0.4)
  * sass (3.7.3)
  * sass-listen (4.0.0)
  * sprockets (3.7.2)
  * thread_safe (0.3.6)
  * typhoeus (1.3.1)
  * tzinfo (1.2.5)
  * yell (2.0.7)
```

**Comment by grauschnabel** [comment]
hi @benswift , at least on your research page the urls ar not clickable for me, so please tell me exactly where to look to solve the problem of this thread.  Thanks very much!

**Comment by benswift** [comment]
I didn’t solve it, sadly. And I tried all the suggestions, so I’m not 100% where to go from here. Sorry.

**Comment by danjjl** [comment]
Has anybody with Jekyll experience figured out how to get the output of the liquid tag to be processed by the markdown engine?

For the moment `{% bibliography --file references %}` produces `..... [http://google.com](http://google.com]`

-> the output of `markdown_links.rb` is not processed by the markdown engine.

**Comment by rawlins** [comment]
Like others the markdown link plugin is not working for me, even though the order of operations documentation implies it should be fine. However, html output is fine, so it wasn't too hard to modify that plugin to generate links that way. The following does what's needed in jekyll 4.0.0 (you can put this as a .rb file in the `_plugins` folder of your site):

```
# Contributed by @mfenner as markdown_links.rb
# See https://github.com/inukshuk/jekyll-scholar/issues/30
# modified to generate html instead of markdown by @rawlins

require 'uri'

module Jekyll
  class Scholar
    class HtmlLinks < BibTeX::Filter
      URL_PATTERN = URI.regexp(['http', 'https', 'ftp'])
      def apply(value)
        value.to_s.gsub(URL_PATTERN) {
          "<a href=\"#{$&}\">#{$&}</a>"
        }
      end
    end
  end
end
```

Does not handle DOIs for the ordering reasons described above, I'm still trying to figure out what would be needed for them for my own purposes.

**Comment by rawlins** [comment]
Ok, here's a solution for DOIs that works for me, though it has some limitations. *First*, if you are using a CSL that inserts URL material into the DOI field, you will need to disable that. For example, the one I'm using had a conditional:

```xml
          <if variable="DOI">
            <text variable="DOI" prefix="DOI: https://doi.org/" /> -->
          </if>
```

so I made a copy of this CSL file in my site directory structure, changed the config so it was using this copy, changed this rule to to:

```xml
          <if variable="DOI">
            <text variable="DOI" prefix="DOI: " />
          </if>
```

*Second*, I installed the following plugin into my site's `_plugins` folder and added it to the list of `bibtex_filters` in the config file:

```ruby
# Convert DOIs to html. This *will not match all DOIs*. The regex comes from
# https://www.crossref.org/blog/dois-and-matching-regular-expressions/
# this may conflict with some CSLs that insert `https://doi.org/` or the like.

require 'uri'

module Jekyll
  class Scholar
    class DOILinks < BibTeX::Filter
      DOI_FILTER = Regexp.new("^10\\.\\d{4,9}/[-._;()/:A-Z0-9]+$", Regexp::IGNORECASE)
      def apply(value)
        value.to_s.gsub(DOI_FILTER) {
          "<a href=\"https://doi.org/#{$&}\">#{$&}</a>"
        }
      end
    end
  end
end
```

As noted, this won't match exotic DOIs (see the link in the code comments above for discussion). For purposes of bibtex value strings, I actually suspect that something much more relaxed like `"^10\\.\\d{3,9}/\\S+$"` might work just fine, as long as you dealt with html escaping. I can't really think of cases where you'd get false positives for this given that it's not trying to match free text.

**Comment by m-pilia** [comment]
The approach mentioned by @benswift did not work for me either, but I think I found a way to go around this issue by writing a custom Liquid filter:

```rb
# put inside file: _plugin/link_to.rb
module Jekyll::CustomFilter
  def link_to(input, href)
    "<a href=\"#{href}\">#{input}</a>"
  end
end
Liquid::Template.register_filter(Jekyll::CustomFilter)
```

Then the `link_to` filter can be used inside the `.cls` file. For instance, I wanted to create an hyperlink in the `title` of the citation whenever a `url` is present, therefore I changed [the relevant part](https://github.com/citation-style-language/styles/blob/bd3a5f6e2/ieee.csl#L132) of the `title` macro as follows:
```xml
<if variable="URL">
    <text variable="title" quotes="true" prefix="{{ '" suffix="' | "/>
    <text variable="URL" prefix="link_to: '" suffix="' }}"/>
</if>
<else>
    <text variable="title" quotes="true"/>
</else>
```
It seems to work, but it feels more like a horrible workaround rather than a proper solution.

**Comment by inukshuk** [comment]
@m-pilia have you tried this in the latest version of jekyll-scholar? I think this will not work anymore, because we just disabled the double parsing of templates.

I think the best option is to use a custom bibliography template; this gives you full access to the entry and to the rendered reference. If you absolutely need the link to be inside the rendered reference, I think the best solution right now is to hack the style, similar to your own example. You can [use this code](https://github.com/inukshuk/jekyll-scholar/issues/286#issuecomment-557697612) to make it possible to insert HTML snippets into the prefix/suffix attributes (using either '<' or '&lt;' should work). (We should probably add something like this to jekyll-scholar behind an option.)

**Comment by m-pilia** [comment]
@inukshuk Thanks for the reply! Indeed, I tried and I can confirm that using Liquid filters inside the CSL does not work on the latetst version. Your suggestion works better though.

**Comment by drscotthawley** [comment]
Has a solution to this ever been made 'standard', or are we still stuck with adding custom hacks?  Latest jekyll-scholar still fails on @x12a1f 's simple test case:

```
@article{test,
      author = {test},
      title = {test},
      URL = {http://google.com}
}
```
Still produces the bibliography entry that reads...
```
test. test. [http://google.com](http://google.com)
```
which is never converted to HTML.  (i.e. no links). 

...this is even with `markdown` and `latex` enabled in config, as per @inukshuk 's instructions:
```
scholar:
  source: _bibliography
  bibtex_filters:
    - markdown
    - latex
```
Turning on markdown filter or not doing so, seems to have no effect.

**Comment by quattrinili** [comment]
I modified the markdown filter to produce the proper HTML, https://github.com/quattrinili/jekyll-scholar/commit/de8cc5dc550549a4c7dc67334614a856e4874d78

I am still testing it to see if that works in general, and probably the right way to do it is to change the name of that filter, but at least it is producing the correct output.

**Comment by ashwin-s-rao** [comment]
I would like to know the steps required to make the title text point the doi. All my bibtex entries have the doi field and I would like to use them. 

I updated my custom csl file as follows

> <macro name="title">
>     <choose>
>         <text variable="title"/>
>         
>         
>         
>     </choose>
>   </macro>
> 

And my _config.yml has the latex and markdown filters enabled. 

> 
> scholar:
>  ...
>   bibtex_filters:
>     - latex
>    
I am using the minima theme and kramdown. 

I tried the html tags `&lt;a href ...` , and even markdown tags `[title](https://dx.doi.org/doi)` but the html pages do not make the titles as links. 

Any suggestions would be useful. I am using jekyll 4.2.0 and jekyll-scholar (6.8.0).

**Comment by AlasdairGray** [comment]
Is there something that needs to be done to get the custom ruby files in `_plugins` to run?

I've been following the instructions of @m-pilia in this [blog](https://martinopilia.com/posts/2020/02/22/migration.html). At the moment I see the HTML string rather than it being interpreted by the browser.
```
doi: <a href="https://doi.org/10.1007/978-3-030-21348-0">10.1007/978-3-030-21348-0</a>
```

**Comment by inukshuk** [comment]
All Ruby files in `_plugins` should get loaded by Jekyll at startup.

Regarding the link, I guess [the solution mentioned in the blog is to downgrade to a previous version of jekyll-scholar](https://martinopilia.com/posts/2020/02/22/migration.html#fn:1). In general, I'd try to remove the DOI from the CSL style and use a custom bibliography template to display the rendered reference and then add your own links for URLs, DOIs and similar.

**Comment by AlasdairGray** [comment]
Thanks, I ended up using the template approach. I hadn't picked up on that in the README. It would be handy to give the set of keys that are available or a pointer to a page that gives them.

**Comment by KonstantinBob** [comment]
> I've been following the instructions of @m-pilia in this [blog](https://martinopilia.com/posts/2020/02/22/migration.html). At the moment I see the HTML string rather than it being interpreted by the browser.
> 
> ```
> doi: <a href="https://doi.org/10.1007/978-3-030-21348-0">10.1007/978-3-030-21348-0</a>
> ```

I was stuck with this, too. Here is a hack that worked at least for my situation:
In my post layout I replaced
```
<div class="post-content" itemprop="articleBody">
 {{ content }}
</div>
```
with 
```
<div class="post-content" itemprop="articleBody">
 {{ content | replace: "&amp;", "&" | replace: "&lt;", "<" | replace: "&gt;", ">" | markdownify }}
</div>
```
and now the links are rendered correctly.

**Comment by amaury-d** [comment]
I had the same issue and found this hopefully elegant enough solution (no plugin)

**_bibliography/references.bib**
```bibtex
@article{test,
  author     = {test},
  title      = {test},
  hidden_url = {http://google.com}
}
```

**_config.yml**
```yaml
scholar:
  bibliography_template: bib
```

**_layouts/bib.html**
```html
---
---

{% if entry.hidden_url %}
<a href="{{ entry.hidden_url }}">{{reference}}</a>
{% else %} {{reference}} {% endif %}
```

**Comment by Sciurus365** [comment]
Hi all, a follow-up from the GPT age. I'm not sure if a similar solution has been mentioned above as I'm not an expert of Ruby, but I discussed GPT and found the following solution that works perfectly with APA 7.

First, create a file called `doi_linkify.rb` in `_plugins`:

```ruby
module Jekyll
  module DOIConverter
    def doi_linkify(input)
      input.gsub(/(https?:\/\/doi\.org\/[^\s"<]+)/i) do |doi_url|
        %(<a href="#{doi_url}">#{doi_url}</a>)
      end
    end
  end
end

Liquid::Template.register_filter(Jekyll::DOIConverter)
```

And then, change the following in `_layouts\page.html`

```html
{{ content }}
```

to

```html
{{ content | doi_linkify }}
```

**Comment by ashinkarov** [comment]
Well, this assumes that:
1. all your urls are of the shape `https://doi.org...`
2. the content of the `<a>???</a>`  that you generate is `https://...`

This might be ok for a particular usecase, but it feels quite restrictive in general.  What if you have a bib file with a bunch of urls some of which you want to show in a certain way?

**Comment by Sciurus365** [comment]
> Well, this assumes that:
> 1. all your urls are of the shape `https://doi.org...`
> 2. the content of the `<a>???</a>`  that you generate is `https://...`
> 
> This might be ok for a particular usecase, but it feels quite restrictive in general.  What if you have a bib file with a bunch of urls some of which you want to show in a certain way?
>    

Hi you are right, I am using this because in my field we have the convention to use doi almost exclusively. I am just sharing this in case some other people have the similar need.
